options.go

```diff
@@ -41,6 +48,48 @@ func CaseInsensitive(tokens ...string) Option {
 	}
 }
 
+func UseCustom(parseFn interface{}) Option {
+	errorType := reflect.TypeOf((*error)(nil)).Elem()
+	peekingLexerType := reflect.TypeOf((*lexer.PeekingLexer)(nil))
+	return func(p *Parser) error {
+		parseFnVal := reflect.ValueOf(parseFn)
+		parseFnType := parseFnVal.Type()
+		if parseFnType.Kind() != reflect.Func {
+			return fmt.Errorf("production parser must be a function (got %s)", parseFnType)
+		}
+		if parseFnType.NumIn() != 1 || parseFnType.In(0) != reflect.TypeOf((*lexer.PeekingLexer)(nil)) {
+			return fmt.Errorf("production parser must take a single parameter of type %s", peekingLexerType)
+		}
+		if parseFnType.NumOut() != 2 {
+			return fmt.Errorf("production parser must return exactly two values: the parsed production, and an error")
+		}
+		if parseFnType.Out(0).Kind() != reflect.Interface {
+			return fmt.Errorf("production parser's first return must be an interface type")
```

Comment:
@alecthomas In the issue thread for #229 we discussed that this does not necessarily need to return an interface type. I'm flip-flopping back and forth on agreement with this.

On the one hand, in support of that conclusion: absolutely nothing about this code _requires_ that this type is an interface. The parse function could return anything at all.

On the other hand, against this conclusion:
What happens if someone provides a parsing function for `bool`, `Position`, `string`, `int64`, or any other type that the parser expects to have a specific meaning? I haven't tested it, but I suspect that this would lead to surprising behavior that could potentially be hard to track down.

It seems like there has to be some kind of limitation on the particular type(s) that can be parsed using a custom parsing function. `interface` types seem like a logical constraint, and I think it could even see that being extended to other complex types like structs, pointers-to-structs, slices-of-structs, and arrays-of-structs... but I'm not sure how to decide on a proper sub-set of types that are allowed.

It is _not_ a breaking change to accept more types, but if we have to remove types that are allowed, that _is_ a breaking change. That might be totally fine, as well - I'm of course open to whatever you think is the right initial approach - its just that my instinct is to be cautious: we could release the first version of this feature supporting only `interface` types, and then expand the allowed types in the future based on feedback from users who need other types here.

Reply:
Good thinking, I agree it's a reasonable place to start.